[D2] 기프티콘 응모 화면 구현

AhnSangHee opened the pull request:
# 개요
- 🔗  이슈링크 : resolve https://github.com/mash-up-kr/GGiriGGiri_iOS/issues/51


# 변경사항
## 💻 작업 내용
- 기프티콘 응모 화면 구현
   - 응모 버튼 누르면 응모 완료 알림창 띄우기
   - 응모 완료되면 응모 완료 버튼으로 변경하기
- 메인 화면에서 선택된 기프티콘의 id 값을 응모 화면으로 전달하도록 구현
   - Gifticon에 id 프로퍼티 추가 → DeadLineCollectionViewCell, GifticonCardCollectionViewCell에도 이에 대응하기 위한 gifticonId 프로퍼티 추가
   - CategoryCell에 categoryType 추가

## 📱 미리보기
- 시간 표시 뷰는 컴포넌트이기 때문에 임시로 빈 뷰로 표시해두었습니다.
- 응모 버튼을 누르면 뜨는 모달 창은 컴포넌트이기 때문에 임시로 알림창이 뜨도록 구현해두었습니다.

![Simulator Screen Recording - iPhone 13 Pro Max - 2022-07-27 at 04 36 20](https://user-images.githubusercontent.com/95578975/181097016-7f655947-c540-4446-855a-d4a4e9adc1a6.gif)

AhnSangHee committed:
:sparkles::bulb: 기프티콘 응모 VC 추가 (#51)

- MainCollectionViewCellDelegate 프로토콜 추가
- 기프티콘 응모 VC에 내비게이션 바 추가

AhnSangHee committed:
:bulb: 기프티콘 응모 화면에 카드뷰 추가 (#51)

AhnSangHee committed:
:sparkles::bulb: 기프티콘 응모 화면 구현 (#51)

- ApplyGifticonView를 ApplyGifticonCardView, ApplyGifticonInfoView, ApplyGifticonDDipInfoView의 세 부분으로 나누어 구현

AhnSangHee committed:
:sparkles: UIButton Extension 추가 (#51)

- state에 따라서 button의 backgroundColor 변경되도록 하는 setBackgroundColor 메서드 추가

AhnSangHee committed:
:bulb: 응모 버튼 누르면 버튼 비활성화되도록 구현 (#51)

AhnSangHee committed:
:bulb::recycle: Gifticon에 기프티콘 고유값으로 쓰일 id 프로퍼티 추가 (#51)

- MockData에도 id 프로퍼티 추가

AhnSangHee committed:
:bulb::recycle: CategoryCell에 categoryType 프로퍼티 추가 (#51)

AhnSangHee committed:
:bulb: ApplyViewModel 추가 (#51)

AhnSangHee committed:
:bulb: MainCollectionViewCell select 이벤트 구현 (#51)

- 마감 임박, 기프티콘 목록 Section은 cell이 갖고 있는 id를 ApplyVM으로 주입함으로써 ApplyVC를 초기화하도록 함
- 카테고리 Section은 터치 이벤트는 정렬이기 때문에 추후 구현 예정

AhnSangHee committed:
:recycle: 기프티콘 이미지뷰 contentMode 변경 (#51)

AhnSangHee committed:
:bulb: 기프티콘 응모 버튼 클릭하면 버튼 텍스트 변경되도록 구현 (#51)

AhnSangHee committed:
:recycle: 코드 컨벤션 수정 (#51)

szzang-kr commented on Projects/App/Sources/Apply/View/ApplyGifticonInfoView.swift:
```diff
@@ -0,0 +1,121 @@
+            $0.top.leading.equalToSuperview()
+        }
+        
+        infoContainerView.snp.makeConstraints {
+            $0.top.equalTo(gifticonInfoSectionTitleLabel.snp.bottom).offset(20)
+            $0.leading.trailing.equalToSuperview()
+            $0.height.equalTo(204)
+        }
```
infoContainerView가 stackView를 subView로 가지고있는 구조인데, height이 204로 고정되어있어 stackView의 height이 204보다 늘어나게 되면 stakView의 내용이 전부 표현되지 않을 수 있습니다.

특별하게 고정한 의도가 있을까요?

AhnSangHee replied:
피그마에 있는 높이 값을 그대로 설정한 것인데요, 말씀해주신 이유로 height 값을 고정하지 않는 것에 동의합니다. height 설정하는 부분을 제거했습니다!

kimkyunghun3 commented on Projects/App/Sources/Apply/ApplyViewController.swift:
```diff
@@ -0,0 +1,71 @@
+    override func setLayout() {
+        super.setLayout()
+        
+        view.addSubviews(with: [applyGifticonView])
```
이부분에는 하나만 넣고 있는데 addSubViews로 한 이유가 있을까요?

AhnSangHee replied:
이전에 다른 뷰도 넣다가 제거하면서 이렇게 되었네요..😅 `addSubview`를 사용하도록 변경했습니다~!

kimkyunghun3 commented on Projects/App/Sources/Apply/View/ApplyGifticonDDipInfoView.swift:
```diff
@@ -0,0 +1,112 @@
+        return view
+    }()
+    
+    private let verticalStackView: UIStackView = {
```
verticalStackView라고 했을 때 어떤 스택뷰인지 유추하기가 어렵습니다..!

szzang-kr replied:
개인적인 의견은 내부에서 제한적으로 사용되는 UI property의 경우 이미 의미가 부모에 포함 되어 있다고 생각하여 간단하게 짓고 있습니다.

```swift
class InformationView: UIView {
  private let stackView: UIStackView
}
```
해당 코드에서는 이미 스택뷰가 어떤 정보를 담고있을지 class name에 명시 되어있다고 생각합니다.

따라서 해당 verticalStackView도 GifticonDDipInfo를 가지고있는 stackView라고 유추 가능 할 것 같습니다.

ios-h replied:
infoContainerView 내부에 있는 단순 verticalStackView인데요, private하게 쓰이기도 하고 말그대로 뷰들을 vertical하게 담는 역할밖에 하지 않기 때문에 어떤 스택뷰인지 네이밍을 할 필요가 없다고 생각합니다. 물론 스택뷰가 여러 개가 생긴다면 더 정확한 네이밍으로 만들어야겠지만, 추후 저희 기획에서에 스택뷰가 추가될 일이 없기 때문에 괜찮다고 생각합니다!

AhnSangHee replied:
infoContainerView 내부에 있는 단순 verticalStackView인데요, private하게 쓰이기도 하고 말그대로 뷰들을 vertical하게 담는 역할밖에 하지 않기 때문에 어떤 스택뷰인지 네이밍을 할 필요가 없다고 생각합니다. 물론 스택뷰가 여러 개가 생긴다면 더 정확한 네이밍으로 만들어야겠지만, 추후 저희 기획에서도 스택뷰가 추가될 일이 없기 때문에 괜찮다고 생각합니다!

kimkyunghun3 replied:
아하 제 생각에는 이 클래스 내부에서 가지고 있기에 물론 유추가 가능하지만 이는 유추라는 말은 즉, 오해를 살수도 있다고 생각했습니다. 그래서 유추보다는 더 명확하게 제공하는 것이 제3자가 오해하지 않고 정확하게 사용할 수 있을 것이라고 생각했습니다.
그래서 말씀해주신 것도 일리가 있는데 제 생각에서는 명확하게 하면 더 좋을거같다는 단순한 의견이였습니다!

kimkyunghun3 commented on Projects/App/Sources/Apply/View/ApplyGifticonDDipInfoView.swift:
```diff
@@ -0,0 +1,112 @@
+        return stackView
+    }()
+    
+    private let categoryContainerView = InfoView()
```
요렇게 간단하게 생성한 것과 클로저로 인스턴스로 생성한 부분이 여러곳 있는데, 저는 보통 간단한 생성은 묶어두고 그 밑에 클로저로 생성한 것으로 묶어두도록 해서 가독성을 높입니다! 이는 참고사항이니 참고만 해주세요 :)

AhnSangHee replied:
오오 저는 시각적으로 봤을 때 위에서부터 차례대로 배치되는 뷰 순서로 선언을 하는 편인데 말씀하신 방법도 좋은 것 같아요 😀

kimkyunghun3 commented on Projects/App/Sources/Apply/View/ApplyGifticonDDipInfoView.swift:
```diff
@@ -0,0 +1,112 @@
+    override func setLayout() {
+        super.setLayout()
+        
+        
```
공백이 두개있어요!!

kimkyunghun3 commented on Projects/App/Sources/Apply/View/ApplyGifticonDDipInfoView.swift:
```diff
@@ -0,0 +1,112 @@
+        super.setLayout()
+        
+        
+        verticalStackView.addArrangedSubviews(with: [categoryContainerView])
```
한가지만 넣고 있는데 스택뷰로 한 이유와 addArrangedSubViews로 한 이유가 있을까요?

AhnSangHee replied:
이전에 다른 뷰도 넣다가 제거하면서 이렇게 되었네요..😅 `addArrangedSubview`를 사용하도록 변경했습니다~!

kimkyunghun3 commented on Projects/App/Sources/Apply/View/ApplyGifticonInfoView.swift:
```diff
@@ -0,0 +1,121 @@
+    override func setLayout() {
+        super.setLayout()
+        
+        
```
여기도 공백이 두개에요!

kimkyunghun3 commented on Projects/App/Sources/Apply/View/ApplyGifticonView.swift:
```diff
@@ -0,0 +1,116 @@
+    
+    public weak var delegate: ApplyGifticonViewButtonDelegate?
+    
+    private let scrollView: UIScrollView = {
```
무슨 스크롤뷰인지 명시적으로 해주시면 좋을거같아요~

ios-h replied:
말 그대로 `ApplyGifticonView` 내에서만 쓰이는 스크롤뷰이기 때문에 단순하게 `scrollView`라고 이름을 지었습니다! 혹시 네이밍을 추천해주실 수 있을까요 😮

ios-h replied:
말 그대로 `ApplyGifticonView` 내에서만 쓰이는 스크롤뷰이기 때문에 단순하게 `scrollView`라고 이름을 지었습니다! 추가로 추후 저희 기획에서도 스크롤뷰가 추가될 일이 없기 때문에 이렇게 했는데요. 혹시 생각하시는 네이밍이 있을까요? 😮

ios-h replied:
말 그대로 `ApplyGifticonView` 내에서만 쓰이는 스크롤뷰이기 때문에 단순하게 `scrollView`라고 이름을 지었습니다! 추가로 추후 저희 기획에서도 스크롤뷰가 추가될 일이 없기 때문에 이렇게 했는데요. 혹시 생각하시는 네이밍이 있을까요? 😮

ios-h replied:
말 그대로 ApplyGifticonView 내에서만 쓰이는 스크롤뷰이기 때문에 단순하게 scrollView라고 이름을 지었습니다. 추가로 추후 저희 기획에서도 스크롤뷰가 추가될 일이 없기 때문에 이렇게 했는데요. 굳이 무슨 스크롤뷰인지 명시할 필요가 없다고 생각합니다! 혹시 생각하시는 네이밍이 있을까요? 😮

AhnSangHee replied:
말 그대로 ApplyGifticonView 내에서만 private하게 쓰이는 스크롤뷰이기 때문에 단순하게 scrollView라고 이름을 지었습니다. 추가로 추후 저희 기획에서도 스크롤뷰가 추가될 일이 없기 때문에 이렇게 했는데요. 굳이 무슨 스크롤뷰인지 명시할 필요가 없다고 생각합니다! 혹시 생각하시는 네이밍이 있을까요? 😮

kimkyunghun3 replied:
여기에서는 일단 스크롤뷰가 하나라서 괜찮을거같긴하네요! 추후 변동이 없다면 냅두어도 괜찮아보이지만, 저는 조금 더 명시적인게 제3자 입장에서 휴먼오류를 발생하는 것을 줄일 수 있다고 생각합니다.
아무래도 다른 사람의 코드를 보는 것이기에 제대로 이해하려고 해도 안되는 경우가 있기에 이는 또 추가적인 소통의 시간이 필요할 것이라고 생각합니다.
그래서 저는 더 명확한 것이 이런 소통을 더 줄일수 있지않을까 생각했던것입니다 ㅎㅎ

kimkyunghun3 commented on Projects/App/Sources/Main/MainCollectionViewDelegate.swift:
```diff
@@ -8,15 +8,31 @@
+}
+
 final class MainCollectionViewDelegate: NSObject, UICollectionViewDelegate {
+    weak var collectionViewCellDelegate: MainCollectionViewCellDelegate?
```
Delegate에 public 이 붙어있는 곳이 있고 안붙어있는 곳이 있는데 차이가 있을까요?

ios-h replied:
없습니다..ㅎㅎ `public`으로 명시하겠습니다~!

AhnSangHee replied:
차이는 없습니다~! 현재 존재하는 모든 delegate가 internal 수준의 접근 제어를 가지면 된다고 생각하기 때문에 delegate에 있는 접근 제어자를 모두 제거했습니다.

kimkyunghun3 commented on Projects/App/Sources/Apply/ViewModel/ApplyViewModel.swift:
```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+protocol ApplyViewModelProtocol {
+    var id: Int { get set }
```
정확하게는 파악이 어려운데 set으로 id값이 바뀌나요?

AhnSangHee replied:
바뀌지 않습니다. set 키워드 제거하였습니다 지적 감사합니다~!

kimkyunghun3 reviewed (changes requested):
고생했습니다!!!

szzang-kr commented on Projects/App/Sources/Apply/ViewModel/ApplyViewModel.swift:
```diff
@@ -0,0 +1,21 @@
+    var id: Int { get set }
+}
+
+final class ApplyViewModel: ApplyViewModelProtocol {
+    var id: Int
+    
+    init(id: Int) {
+        self.id = id
+    }
+}
```
ApplyViewModel의 네이밍으로 미루어보아 신청하려는 항목의 id로 유추가 됩니다.

하지만 ApplyViewController 에서는 viewModel.id를 다시 gifticonId로 할당하여 사용합니다.

이렇게되면 viewModel의 id는 어느 용도로 사용되는지 ViewController를 보지 않는다면 알 수 없습니다.

명확하게 `gifticonId`로 propertyname을 정의하는쪽이 협업에 있어서 커뮤니케이션 비용을 줄일 수 있을것 같습니다. 😓

AhnSangHee replied:
코멘트 달아주신 내용에 동의합니다. 명확하게 id에서 gifticonId로 변경하였습니다!😀

szzang-kr commented on Projects/App/Sources/Common/Extension/UIButton+.swift:
```diff
@@ -0,0 +1,26 @@
+    func setBackgroundColor(_ color: UIColor?, for state: UIControl.State) {
+        
+        UIGraphicsBeginImageContext(CGSize(width: 1.0, height: 1.0))
+        guard let context = UIGraphicsGetCurrentContext() else { return }
+        guard let color = color else { return }
```
```suggestion
        guard 
            let context = UIGraphicsGetCurrentContext(),
            let color = color
        else { return }
```
위와 같이 한줄에 조건을 표시할 수도 있습니다. 분리하신 이유가 있을까요 ? 😄

AhnSangHee replied:
처음에는 color 조건이 없었는데, 추후 추가하면서 저렇게 되었네요😅 한 줄에 조건을 모두 표시하도록 [변경](https://github.com/mash-up-kr/GGiriGGiri_iOS/pull/72/commits/4bd65e6415eafd9474bdadaf15dafaebd336d43e)했습니다!

szzang-kr commented on Projects/App/Sources/Main/MainViewController.swift:
```diff
@@ -130,6 +134,18 @@ final class MainViewController: BaseViewController<MainViewModelProtocol> {
         picker.delegate = self
         present(picker, animated: true)
     }
+    
+    func gifticonCellTapped(with id: Int) {
+        let applyViewModel = ApplyViewModel(id: id)
+        let applyViewController = ApplyViewController(applyViewModel)
+        applyViewController.modalPresentationStyle = .fullScreen
+        self.present(applyViewController, animated: true)
+    }
+    
+    func categoryCellTapped(with category: Category) {
+        // TODO: Category에 따라 정렬하기
+        debugPrint(#function)
+    }
```
🟡 delegate method들은 extension으로 모아두는것을 제안드립니다.

MainViewController에 정의된 method와 같은 scope에 있으면 MainViewController를 위한 method인지, delegate method인지 구분하기 힘들기 때문입니다.

추가로 delegate method만 모아볼 수 있는 장점도 있습니다. 😄

AhnSangHee replied:
extension으로 빼는게 훨씬 더 좋은 것 같네요!! [여기](https://github.com/mash-up-kr/GGiriGGiri_iOS/pull/72/commits/f8fde695a85d74580c1b08f9608f3cb8c544ec52)에 반영하였습니다 😀

szzang-kr reviewed (approved):
수고하셨습니다 ~!! 👍

AhnSangHee committed:
:recycle: 코드 컨벤션 수정 (#51)

AhnSangHee committed:
:recycle: 코드 컨벤션 수정 (#51)

AhnSangHee committed:
:recycle: delegate 메서드를 extension으로 분리 (#51)

AhnSangHee committed:
:recycle: 코드 컨벤션 수정 및 접근 제어자 변경 (#51)

AhnSangHee committed:
:recycle: ApplyVM, ApplyGifticonInfoView 수정  (#51)

- ApplyVM id → gifticonId 로 변경 및 set 키워드 제거
- ApplyGifticonInfoView 내부의 infoContainerView height 고정값 제거
